File: CPP/Binary Search/BinarySearchCircularRotation.cpp
//How many times a circular array is rotated? 
#include <iostream>
#include <vector>

using namespace std;

int find_rotation_count(vector<int> v)
{
	int low = 0;
	int n = v.size();
	int high = n-1;
	int mid = 0;
	while(high >= low)
	{
		mid = low + (high - low)/2;
		
		if(v[low] <= v[high])
			return low;
		int next = (mid+1)%n;
		int prev = (mid-1+n)%n;
		
		if(v[mid] <= v[next] && v[mid] <= v[prev])
			return mid;
			
		if(v[mid] <= v[high])
			high = mid - 1;
		if(v[mid] >= v[low])
			low = mid + 1;
	}
	return -1;
}

int main()
{
	int n;
	cin>>n;
	vector<int> v(n,0);
	for(int i=0;i<n;i++)
	{
		cin>>v[i];
	}
	int result = find_rotation_count(v);
	if(result != -1)
		cout<<"The array is rotated "<<result<<" times "<<endl;
	
	return 0;
}

File: Java/Binary Search/BinarySearchCircularRotation.java
import java.util.Scanner;

public class BinarySearchCircularRotation {
    
    private static int findRotationCount(int input[], int n)
    {
        int start = 0;
        int end = n-1;
        while(start <= end)
        {
            int mid = start + (end - start)/2;
            if(input[start] <= input[end])
            {
                return start;
            }
            int next = (mid + 1)%n;
            int prev = (mid - 1 + n)%n;
            if(input[mid] <= input[next] && input[mid] <= input[prev])
            {
                return mid;
            } 
            else if(input[mid] <= input[end])
            {
                end = mid - 1;
            }
            else if(input[start] <= input[mid])
            {
                start = mid + 1;
            }
        }
        return -1;
    }

    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int input[] = new int[n];
        for(int i=0;i<n;i++)
        {
            input[i] = sc.nextInt();
        }
        System.out.println("Circular Array Rotation Count: " + findRotationCount(input, n));
        sc.close();
    }
}
